File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-apps/src/state/appsState.ts
import type { AppsPageResult } from "../types/appsModels.js";

export interface AppsState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: AppsPageResult<unknown>;
  readonly error?: string;
}

export function initialAppsState(): AppsState {
  return { status: "idle" };
}

export function reduceAppsState(
  state: AppsState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: AppsPageResult<unknown> }
    | { type: "failed"; error: string },
): AppsState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}

File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_apps/lib/src/state/apps_state.dart
import '../models/apps_models.dart';

enum AppsStatus { idle, loading, ready, error }

class AppsState {
  const AppsState({
    this.status = AppsStatus.idle,
    this.lastResult,
    this.error,
  });

  final AppsStatus status;
  final AppsPageResult<Object?>? lastResult;
  final String? error;

  AppsState loading() => const AppsState(status: AppsStatus.loading);

  AppsState loaded(AppsPageResult<Object?> result) =>
      AppsState(status: AppsStatus.ready, lastResult: result);

  AppsState failed(String message) =>
      AppsState(status: AppsStatus.error, error: message);
}
